kafka-rest/src/main/java/io/confluent/kafkarest/ProducerMetrics.java

```diff
@@ -0,0 +1,414 @@
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.WindowedCount;
+import org.apache.kafka.common.metrics.stats.WindowedSum;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProducerMetrics {
+  public static final String JMX_PREFIX = "kafka.rest";
+
+  private static final Logger log = LoggerFactory.getLogger(ProducerMetrics.class);
+
+  private final Metrics metrics;
+  private final ConcurrentMap<BeanCoordinate, ProduceMetricMBean> beansByCoordinate =
+      new ConcurrentHashMap<>();
+
+  public ProducerMetrics(Time time) {
+    this.metrics = new MetricsBuilder(JMX_PREFIX).withTime(time).build();
+    setupMetricBeans();
+    log.info("Successfully registered kafka-rest produce metrics with JMX");
+  }
+
+  /**
+   * Get or create a {@link ProduceMetricMBean} with the specified name and the given tags. Each
+   * group is uniquely identified by the name and tags.
+   *
+   * @param groupName the name of the ReplicatorMetricGroup group; may not be null
+   * @param tags pairs of tag name and values
+   * @return the {@link ProduceMetricMBean} that can be used to create metrics; never null
+   */
+  public ProduceMetricMBean mbean(String groupName, Map<String, String> tags) {
+    BeanCoordinate beanCoordinate = new BeanCoordinate(groupName, tags);
+    beansByCoordinate.putIfAbsent(beanCoordinate, new ProduceMetricMBean(beanCoordinate));
+    return beansByCoordinate.get(beanCoordinate);
+  }
+
+  /** Stop and unregister the metrics from any reporters. */
+  public void stop() {
+    log.info("Unregistering produce metrics with JMX");
+    metrics.close();
+  }
+
+  /** Sets up a {@link ProduceMetricMBean} for each sensor */
+  private void setupMetricBeans() {
+    ProduceMetricMBean metricMBean =
+        mbean(ProducerMetricsRegistry.GROUP_NAME, Collections.emptyMap());
+
+    // Clear out any previous metrics and sensors that might have been present
+    if (metricMBean != null) {
+      metricMBean.close();
+    }
+
+    // request metrics
+    new SensorBuilder(metricMBean, ProducerMetricsRegistry.REQUEST_SENSOR)
+        .addRate(ProducerMetricsRegistry.REQUEST_RATE, ProducerMetricsRegistry.REQUEST_RATE_DOC)
+        .addCumulativeCount(
+            ProducerMetricsRegistry.REQUEST_TOTAL, ProducerMetricsRegistry.REQUEST_TOTAL_DOC)
+        .addWindowedCount(
+            ProducerMetricsRegistry.REQUEST_TOTAL_WINDOWED,
+            ProducerMetricsRegistry.REQUEST_TOTAL_WINDOWED_DOC)
+        .build();
+
+    new SensorBuilder(metricMBean, ProducerMetricsRegistry.REQUEST_SIZE_SENSOR)
+        .addAvg(
+            ProducerMetricsRegistry.REQUEST_SIZE_AVG, ProducerMetricsRegistry.REQUEST_SIZE_AVG_DOC)
+        .addCumulativeSum(
+            ProducerMetricsRegistry.REQUEST_SIZE_TOTAL,
+            ProducerMetricsRegistry.REQUEST_SIZE_TOTAL_DOC)
+        .addWindowedSum(
```

Comment:
Should this be addWindowedCount rather than the WindowedSum (I'm assuming a Windowed sum is a running total of all the separate windowedCounts so far)?  

The changes in rest-utils didn't add any counter based metrics to the sizes (just the count of requests and count of errors), so I guess it depends on what we think adding up all the sizes of the requests so far will tell us, and if that's useful.

Reply:
these were mirroring the topic level bytes metrics from the other clients but I can't think of a major use for them for cloud monitoring so have removed.